shell/platform/embedder/embedder.cc

```diff
@@ -712,6 +712,25 @@ inline flutter::PointerData::SignalKind ToPointerDataSignalKind(
   return flutter::PointerData::SignalKind::kNone;
 }
 
+// Returns the buttons for PointerData for the given buttons and change from a
+// FlutterPointerEvent.
+inline int64_t ToPointerDataButtons(
+    int64_t buttons,
+    flutter::PointerData::Change change) {
+  // `buttons` should be non-zero when and only when the pointer is down.
```

Comment:
Is the "and only when" actually true? For instance, is sending a `cancel` with a button set actually an error? I can think of cases where I would expect that to be the event created by the platform on desktop, unless it were explicitly worked around at the platform level.

Reply:
Cancel is kind of special being the state of "termination". I can add it to the list (down, move or cancel), but other events should really not involve buttons.

For other events, there is a set of restrictions we hope the list of `PointerData` should obey, and the engine should regularize them before sending to framework. It is currently handled by [`PointerEventConverter `](https://github.com/flutter/flutter/blob/7beb09e76af70b92a6c03879fb22eed1ab86a8c3/packages/flutter/lib/src/gestures/converter.dart#L103). There was [a plan](https://github.com/flutter/engine/pull/8064) to move them to the engine (leaving framework with only 1-to-1 converting) but was cancelled. I'm not sure we're still going this way though.